File: src/com/alchitry/labs/gui/main/MainMenu.java
package com.alchitry.labs.gui.main;

import java.io.File;
import java.io.IOException;

import org.eclipse.swt.SWT;
import org.eclipse.swt.events.SelectionAdapter;
import org.eclipse.swt.events.SelectionEvent;
import org.eclipse.swt.widgets.FileDialog;
import org.eclipse.swt.widgets.Menu;
import org.eclipse.swt.widgets.MenuItem;
import org.eclipse.swt.widgets.Shell;

import com.alchitry.labs.Reporter;
import com.alchitry.labs.Settings;
import com.alchitry.labs.Util;
import com.alchitry.labs.gui.BoardSelector;
import com.alchitry.labs.gui.EmailMessage;
import com.alchitry.labs.gui.FeedbackDialog;
import com.alchitry.labs.gui.NewProjectDialog;
import com.alchitry.labs.gui.StyledCodeEditor;
import com.alchitry.labs.gui.ThemeSelectorDialog;
import com.alchitry.labs.gui.WelcomeDialog;
import com.alchitry.labs.gui.tools.ImageCapture;
import com.alchitry.labs.gui.tools.RegInterface;
import com.alchitry.labs.gui.tools.SerialMonitor;
import com.alchitry.labs.hardware.boards.Board;
import com.alchitry.labs.hardware.flashers.AuFlasher;
import com.alchitry.labs.hardware.flashers.CuFlasher;
import com.alchitry.labs.hardware.flashers.MojoFlasher;
import com.alchitry.labs.project.Project;

public class MainMenu {
	MainWindow parent;
	Menu menu;
	Board board;
	MenuItem iceCube;
	MenuItem iceStorm;

	public MainMenu(MainWindow parent) {
		this.parent = parent;
	}

	public void build() {
		if (menu != null)
			menu.dispose();

		Shell shell = parent.getShell();
		menu = new Menu(shell, SWT.BAR);
		shell.setMenuBar(menu);

		Project p = MainWindow.project;
		if (p != null)
			board = p.getBoard();
		else
			board = null;

		buildFileMenu();
		buildProjectMenu();
		buildToolsMenu();
		buildSettingsMenu();
		buildHelpMenu();
	}

	private MenuItem createItem(Menu m, int style, String text, SelectionAdapter selectedListener) {
		MenuItem menuItem = new MenuItem(m, style);
		menuItem.addSelectionListener(selectedListener);
		menuItem.setText(text);
		return menuItem;
	}

	private MenuItem createCheckItem(Menu m, String text, SelectionAdapter selectedListener) {
		return createItem(m, SWT.CHECK, text, selectedListener);
	}

	private MenuItem createItem(Menu m, String text, SelectionAdapter selectedListener) {
		return createItem(m, SWT.NONE, text, selectedListener);
	}

	private Menu createSubMenu(Menu m, String text) {
		MenuItem menuItem = new MenuItem(m, SWT.CASCADE);
		menuItem.setText(text);

		Menu subMenu = new Menu(menuItem);
		menuItem.setMenu(subMenu);

		return subMenu;
	}

	private void buildFileMenu() {
		Menu subMenu = createSubMenu(menu, "File");
		
		createItem(subMenu, "New File...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				parent.addNewFile();
			}
		});
		
		createItem(subMenu, "Open File...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				FileDialog dialog = new FileDialog(parent.getShell(), SWT.MULTI);
				dialog.setFilterExtensions(new String[] { "*.luc", "*.v", "*" });
				String path = dialog.open();
				if (path != null) {
					String[] files = dialog.getFileNames();
					String dir = dialog.getFilterPath();
					for (String f : files)
						parent.openFile(dir + File.separator + f, true);
				}
			}
		});
		
		createItem(subMenu, "Import Files...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				if (MainWindow.getOpenProject() == null) {
					Util.showError("A project must be open to import a file!");
					return;
				}
				FileDialog dialog = new FileDialog(parent.getShell(), SWT.MULTI);
				dialog.setFilterExtensions(new String[] { "*.luc;*.v;*.ucf", ".pcf", "*" });
				String path = dialog.open();
				if (path != null) {
					String[] files = dialog.getFileNames();
					String dir = dialog.getFilterPath();
					for (String f : files) {
						if (!MainWindow.getOpenProject().importFile(dir + File.separator + f))
							Util.showError("Failed to import \"" + f + "\"");
					}
					MainWindow.getOpenProject().updateTree();
				}
			}
		});
		
		createItem(subMenu, "Save", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				((StyledCodeEditor) (parent.tabFolder.getSelectedControl())).save();
			}
		});

		createItem(subMenu, "New Project...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				parent.createNewProject();
			}
		});

		createItem(subMenu, "Open Project...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				parent.openProject();
			}
		});

		createItem(subMenu, "Clone Project...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				if (MainWindow.project == null || !MainWindow.project.isOpen()) {
					Util.showError("You need to open or create a project first!");
					return;
				}
				NewProjectDialog dialog = new NewProjectDialog(parent.getShell(), SWT.DIALOG_TRIM, true);
				parent.getShell().setEnabled(false);
				Project p = dialog.open();
				if (p != null) {
					try {
						if (MainWindow.project.saveAsXML(p.getProjectFolder(), p.getProjectName()))
							parent.openProject(p.getProjectFolder() + File.separatorChar + p.getProjectName() + ".alp");
						else
							Util.showError("Failed to clone project! Project folder was null!");
					} catch (IOException e1) {
						Util.showError("Failed to clone project!");
						e1.printStackTrace();
					}
				}
				parent.getShell().setEnabled(true);
			}
		});

		createItem(subMenu, "Exit", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				parent.close();
			}
		});
	}

	private void buildProjectMenu() {
		Menu subMenu = createSubMenu(menu, "Project");

		createItem(subMenu, "Add Components...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				if (MainWindow.project == null || !MainWindow.project.isOpen()) {
					Util.showError("You need to open or create a project first!");
					return;
				}
				parent.addComponents();
			}
		});

		if (Board.isType(board, Board.MOJO))
			createItem(subMenu, "Launch CoreGen", new SelectionAdapter() {
				@Override
				public void widgetSelected(SelectionEvent e) {
					if (MainWindow.project == null || !MainWindow.project.isOpen()) {
						Util.showError("You need to open or create a project first!");
						return;
					}
					parent.coreGen.launch(MainWindow.project);
				}
			});
	}

	private void buildToolsMenu() {
		Menu subMenu = createSubMenu(menu, "Tools");

		createItem(subMenu, "Flash Firmware...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				BoardSelector selector = new BoardSelector(parent.getShell(), SWT.APPLICATION_MODAL);
				Board b = selector.open();
				if (b == null)
					return;

				switch (b.getType()) {
				case Board.AU:
					new AuFlasher().flash();
					break;
				case Board.CU:
					new CuFlasher().flash();
					break;
				case Board.MOJO:
					new MojoFlasher().flash();
					break;
				}
			}
		});

		createItem(subMenu, "Serial Port Monitor...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				if (parent.monitor == null || parent.monitor.isDisposed()) {
					if (MainWindow.project.isBusy()) {
						Util.showError("Operation already in progress!");
					} else {
						if (!parent.display.isDisposed())
							parent.monitor = new SerialMonitor(parent.display);
					}
				} else {
					parent.monitor.setFocus();
				}
			}
		});

		createItem(subMenu, "Register Interface...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				if (parent.display.isDisposed())
					return;

				if (parent.regInterface == null || parent.regInterface.isDisposed()) {
					parent.regInterface = new RegInterface(parent.display);

				} else {
					parent.regInterface.setFocus();
				}
			}
		});

		createItem(subMenu, "Image Capture...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				if (parent.imgCapture == null || parent.imgCapture.isDisposed()) {
					parent.imgCapture = new ImageCapture(parent.display);

				} else {
					parent.imgCapture.setFocus();
				}
			}
		});

		if (Board.isType(board, Board.MOJO))
			createItem(subMenu, "Wave Capture...", new SelectionAdapter() {
				@Override
				public void widgetSelected(SelectionEvent e) {
					parent.openWave();
				}
			});
	}

	private void buildSettingsMenu() {
		Menu subMenu = createSubMenu(menu, "Settings");

		if (Board.isType(board, Board.MOJO))
			createItem(subMenu, "Serial Port...", new SelectionAdapter() {
				@Override
				public void widgetSelected(SelectionEvent e) {
					parent.selectSerialPort();
				}
			});

		createItem(subMenu, "ISE Location...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				parent.updateISELocation();
			}
		});

		createItem(subMenu, "Vivado Location...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				parent.updateVivadoLocation();
			}
		});

		createItem(subMenu, "iCEcube2 Location...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				parent.updateIcecubeLocation();
			}
		});

		createItem(subMenu, "iCEcube2 License Location...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				parent.updateIcecubeLicenseLocation();
			}
		});

		createItem(subMenu, "IceStorm Locations...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				parent.updateYosysLocation();
				parent.updateArachneLocation();
				parent.updateIcepackLocation();
			}
		});

		Menu cuBuilder = createSubMenu(subMenu, "Cu Toolchain");

		iceCube = createCheckItem(cuBuilder, "iCEcube2", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				Settings.pref.putBoolean(Settings.USE_ICESTORM, false);
				updateCuToolchainSelection();
			}
		});

		iceStorm = createCheckItem(cuBuilder, "IceStorm", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				Settings.pref.putBoolean(Settings.USE_ICESTORM, true);
				updateCuToolchainSelection();
			}
		});

		updateCuToolchainSelection();

		createItem(subMenu, "Choose Theme...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				ThemeSelectorDialog dialog = new ThemeSelectorDialog(parent.shlAlchitryLabs);
				dialog.open();
			}
		});
	}

	private void updateCuToolchainSelection() {
		boolean useStorm = Settings.pref.getBoolean(Settings.USE_ICESTORM, false);
		iceCube.setSelection(!useStorm);
		iceStorm.setSelection(useStorm);
	}

	private void buildHelpMenu() {
		Menu subMenu = createSubMenu(menu, "Help");

		createItem(subMenu, "Send Feedback...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				FeedbackDialog feedbackDialog = new FeedbackDialog(parent.shlAlchitryLabs, SWT.APPLICATION_MODAL | SWT.DIALOG_TRIM | SWT.CLOSE);
				EmailMessage message = feedbackDialog.open();
				if (message != null) {
					Reporter.sendFeedback(message);
				}
			}
		});

		createItem(subMenu, "About...", new SelectionAdapter() {
			@Override
			public void widgetSelected(SelectionEvent e) {
				new WelcomeDialog(parent.getShell(), SWT.APPLICATION_MODAL).open();
			}
		});
	}
}

File: src/com/alchitry/labs/gui/main/MainMenu.kt
package com.alchitry.labs.gui.main

import com.alchitry.labs.Settings
import com.alchitry.labs.UpdateChecker
import com.alchitry.labs.Util
import com.alchitry.labs.Util.assembleFile
import com.alchitry.labs.Util.reportException
import com.alchitry.labs.Util.showError
import com.alchitry.labs.gui.*
import com.alchitry.labs.gui.tools.ImageCapture
import com.alchitry.labs.gui.tools.RegInterface
import com.alchitry.labs.gui.tools.SerialMonitor
import com.alchitry.labs.hardware.boards.Board
import org.eclipse.swt.SWT
import org.eclipse.swt.events.SelectionAdapter
import org.eclipse.swt.events.SelectionEvent
import org.eclipse.swt.widgets.FileDialog
import org.eclipse.swt.widgets.Menu
import org.eclipse.swt.widgets.MenuItem
import java.io.IOException

class MainMenu {
    companion object {
        fun createSelectionAdapter(block: (SelectionEvent) -> Unit): SelectionAdapter {
            return object : SelectionAdapter() {
                override fun widgetSelected(e: SelectionEvent) {
                    block.invoke(e)
                }
            }
        }
    }

    private var menu: Menu? = null
    private var board: Board? = null
    private var iceCube: MenuItem? = null
    private var iceStorm: MenuItem? = null

    fun build() {
        menu?.dispose()
        menu = Menu(MainWindow.shell, SWT.BAR)
        MainWindow.shell.menuBar = menu
        board = MainWindow.project?.board
        buildFileMenu()
        buildProjectMenu()
        buildToolsMenu()
        buildSettingsMenu()
        buildHelpMenu()
    }

    private fun createItem(m: Menu, style: Int, text: String, selectedListener: SelectionAdapter): MenuItem {
        val menuItem = MenuItem(m, style)
        menuItem.addSelectionListener(selectedListener)
        menuItem.text = text
        return menuItem
    }

    private fun createCheckItem(m: Menu, text: String, selectedListener: SelectionAdapter): MenuItem {
        return createItem(m, SWT.CHECK, text, selectedListener)
    }

    private fun createItem(m: Menu, text: String, selectedListener: SelectionAdapter): MenuItem {
        return createItem(m, SWT.NONE, text, selectedListener)
    }

    private fun createSubMenu(m: Menu?, text: String): Menu {
        val menuItem = MenuItem(m, SWT.CASCADE)
        menuItem.text = text
        val subMenu = Menu(menuItem)
        menuItem.menu = subMenu
        return subMenu
    }

    private fun buildFileMenu() {
        val subMenu = createSubMenu(menu, "File")
        createItem(subMenu, "New File", createSelectionAdapter { MainWindow.addNewFile() })
        createItem(subMenu, "Open File", createSelectionAdapter {
            val dialog = FileDialog(MainWindow.shell, SWT.MULTI)
            dialog.filterExtensions = arrayOf("*.luc", "*.v", "*")
            val path = dialog.open()
            if (path != null) {
                val files = dialog.fileNames
                val dir = dialog.filterPath
                for (f in files) MainWindow.openFile(assembleFile(dir, f), true)
            }
        })
        createItem(subMenu, "Import Files", createSelectionAdapter {
            if (MainWindow.project == null) {
                showError("A project must be open to import a file!")
                return@createSelectionAdapter
            }
            val dialog = FileDialog(MainWindow.shell, SWT.MULTI)
            dialog.filterExtensions = arrayOf(Util.allFileSuffixes.joinToString(";") { "*$it" }, "*")
            val path = dialog.open()
            if (path != null) {
                val files = dialog.fileNames
                val dir = dialog.filterPath
                for (f in files) {
                    if (MainWindow.project?.importFile(assembleFile(dir, f)) == null) showError("Failed to import \"$f\"")
                }
                MainWindow.project?.updateTree()
            }
        })
        createItem(subMenu, "Save", createSelectionAdapter { (MainWindow.tabFolder.selectedControl as StyledCodeEditor).save() })
        createItem(subMenu, "Print\tCtrl+P", createSelectionAdapter { MainWindow.printOpen() })
        createItem(subMenu, "Exit", createSelectionAdapter { MainWindow.close() })
    }

    private fun buildProjectMenu() {
        val subMenu = createSubMenu(menu, "Project")
        createItem(subMenu, "New Project", createSelectionAdapter {
            MainWindow.createNewProject()
        })
        createItem(subMenu, "Open Project", createSelectionAdapter {
            MainWindow.openProjectDialog()
        })
        createItem(subMenu, "Clone Project", createSelectionAdapter {
            MainWindow.project.let { project ->
                if (project == null) {
                    showError("You need to open or create a project first!")
                    return@createSelectionAdapter
                }
                val dialog = NewProjectDialog(MainWindow.shell, clone = project)
                MainWindow.shell.isEnabled = false
                dialog.open()?.let { MainWindow.openProject(it) }
                MainWindow.shell.isEnabled = true
            }
        })
        createItem(subMenu, "Add Components", createSelectionAdapter {
            if (MainWindow.project == null) {
                showError("You need to open or create a project first!")
                return@createSelectionAdapter
            }
            MainWindow.addComponents()
        })
        if (Board.isType(board, Board.MOJO)) createItem(subMenu, "Launch CoreGen", createSelectionAdapter {
            if (MainWindow.project == null) {
                showError("You need to open or create a project first!")
                return@createSelectionAdapter
            }
            MainWindow.coreGen.launch(MainWindow.project)
        })
        if (Board.isType(board, Board.AU or Board.AU_PLUS)) createItem(subMenu, "Vivado IP Catalog", createSelectionAdapter {
            MainWindow.project.let { project ->
                if (project == null) {
                    showError("You need to open or create a project first!")
                    return@createSelectionAdapter
                }
                MainWindow.vivadoIP.launch(project)
            }
        })
        if (Board.isType(board, Board.AU or Board.AU_PLUS)) createItem(subMenu, "Add Memory Controller", createSelectionAdapter {
            MainWindow.project.let { project ->
                if (project == null) {
                    showError("You need to open or create a project first!")
                    return@createSelectionAdapter
                }
                try {
                    MainWindow.vivadoIP.generateMigCore(project)
                } catch (e1: InterruptedException) {
                    reportException(e1, "Failed to generate MIG core!")
                    showError("Failed to generate MIG core!")
                } catch (e1: IOException) {
                    reportException(e1, "Failed to generate MIG core!")
                    showError("Failed to generate MIG core!")
                }
            }
        })
    }

    private fun buildToolsMenu() {
        val subMenu = createSubMenu(menu, "Tools")
        createItem(subMenu, "Serial Port Monitor", createSelectionAdapter {
            MainWindow.monitor.let { monitor ->
                if (monitor?.isDisposed != false) {
                    if (MainWindow.project?.isBusy == true) {
                        showError("Operation already in progress!")
                    } else {
                        if (!MainWindow.display.isDisposed) MainWindow.monitor = SerialMonitor(MainWindow.display)
                    }
                } else {
                    monitor.setFocus()
                }
            }
        })
        createItem(subMenu, "Register Interface", createSelectionAdapter {
            MainWindow.regInterface.let { regInterface ->
                if (MainWindow.display.isDisposed) return@createSelectionAdapter
                if (regInterface?.isDisposed != false) {
                    MainWindow.regInterface = RegInterface(MainWindow.shell)
                } else {
                    regInterface.setFocus()
                }
            }
        })
        createItem(subMenu, "Image Capture", createSelectionAdapter {
            MainWindow.imgCapture.let { imgCapture ->
                if (imgCapture?.isDisposed != false) {
                    MainWindow.imgCapture = ImageCapture(MainWindow.display)
                } else {
                    imgCapture.setFocus()
                }
            }
        })
        if (Board.isType(board, Board.MOJO or Board.AU or Board.AU_PLUS)) createItem(subMenu, "Wave Capture", createSelectionAdapter {
            MainWindow.openWave()
        })
    }

    private fun buildSettingsMenu() {
        val subMenu = createSubMenu(menu, "Settings")
        createItem(subMenu, "ISE Location", createSelectionAdapter {
            MainWindow.updateISELocation()
        })
        createItem(subMenu, "Vivado Location", createSelectionAdapter {
            MainWindow.updateVivadoLocation()
        })
        createItem(subMenu, "iCEcube2 Location", createSelectionAdapter {
            MainWindow.updateIcecubeLocation()
        })
        createItem(subMenu, "iCEcube2 License Location", createSelectionAdapter {
            MainWindow.updateIcecubeLicenseLocation()
        })
        createItem(subMenu, "IceStorm Locations", createSelectionAdapter {
            MainWindow.updateYosysLocation()
            MainWindow.updateArachneLocation()
            MainWindow.updateIcepackLocation()
        })
        val cuBuilder = createSubMenu(subMenu, "Cu Toolchain")
        iceCube = createCheckItem(cuBuilder, "iCEcube2", createSelectionAdapter {
            Settings.USE_ICESTORM = false
            updateCheckMenu(cuBuilder.items, it.widget as MenuItem)
        })
        iceStorm = createCheckItem(cuBuilder, "IceStorm", createSelectionAdapter {
            Settings.USE_ICESTORM = true
            updateCheckMenu(cuBuilder.items, it.widget as MenuItem)
        })
        updateCuToolchainSelection()
        createItem(subMenu, "Choose Theme", createSelectionAdapter {
            val dialog = ThemeSelectorDialog(MainWindow.shell)
            dialog.open()
        })
        val fontMenu = createSubMenu(subMenu, "Editor Font Size")
        listOf(0.75f, 1.0f, 1.25f, 1.5f, 1.75f, 2.0f).forEach { scale ->
            createCheckItem(fontMenu, "${scale}x", createSelectionAdapter {
                Settings.FONT_SCALE = scale
                Util.showInfo("Restart the IDE to update the font size.")
                updateCheckMenu(fontMenu.items, it.widget as MenuItem)
            })
        }
        val fontPercent = Settings.FONT_SCALE
        for (mi in fontMenu.items) {
            val mip = mi.text.substring(0, mi.text.length - 1).toFloat()
            mi.selection = mip == fontPercent
        }

        val updateMenu = createSubMenu(subMenu, "Check for Updates")
        createCheckItem(updateMenu, "Never", createSelectionAdapter {
            Settings.CHECK_FOR_UPDATES = false
            updateCheckMenu(updateMenu.items, it.widget as MenuItem)
        }).also { it.selection = !Settings.CHECK_FOR_UPDATES }
        createCheckItem(updateMenu, "Stable", createSelectionAdapter {
            Settings.BETA_UPDATES = false
            Settings.CHECK_FOR_UPDATES = true
            updateCheckMenu(updateMenu.items, it.widget as MenuItem)
            UpdateChecker.checkForUpdates()
        }).also { it.selection = Settings.CHECK_FOR_UPDATES && !Settings.BETA_UPDATES }
        createCheckItem(updateMenu, "Beta", createSelectionAdapter {
            Settings.BETA_UPDATES = true
            Settings.CHECK_FOR_UPDATES = true
            updateCheckMenu(updateMenu.items, it.widget as MenuItem)
            UpdateChecker.checkForUpdates()
        }).also { it.selection = Settings.CHECK_FOR_UPDATES && Settings.BETA_UPDATES }

        val reportingMenu = createSubMenu(subMenu, "Anonymous Error Reporting")
        createCheckItem(reportingMenu, "Enable", createSelectionAdapter {
            Settings.ERROR_REPORTING = true
            updateCheckMenu(reportingMenu.items, it.widget as MenuItem)
        }).also { it.selection = Settings.ERROR_REPORTING }
        createCheckItem(reportingMenu, "Disable", createSelectionAdapter {
            Settings.ERROR_REPORTING = false
            updateCheckMenu(reportingMenu.items, it.widget as MenuItem)
        }).also { it.selection = !Settings.ERROR_REPORTING }

        /*
		
		if (Board.isType(board, Board.MOJO)) {
			MenuItem menuItem = new MenuItem(subMenu, SWT.CASCADE);
			menuItem.setText("Serial Port");

			Menu portMenu = new Menu(menuItem);
			menuItem.setMenu(portMenu);
			
			SelectionAdapter portSelectedAdapter = new SelectionAdapter() {
				@Override
				public void widgetSelected(SelectionEvent e) {
					Settings.pref.put(Settings.SERIAL_PORT, ((MenuItem) e.widget).getText());
				}
			};

			menuItem.addArmListener(new ArmListener() {

				@Override
				public void widgetArmed(ArmEvent e) {
					for (MenuItem mi : portMenu.getItems()) 
						mi.dispose();
					for (SerialPort p : SerialPort.getCommPorts()) 
						createCheckItem(portMenu, p.getDescriptivePortName(), portSelectedAdapter);
					String currentPort = Settings.pref.get(Settings.SERIAL_PORT, null);
					for (MenuItem mi : portMenu.getItems()) 
						mi.setSelection(mi.getText().equals(currentPort));
				}
			});
		}
		
		*/
    }

    private fun updateCheckMenu(items: Array<MenuItem>, selectedItem: MenuItem) {
        for (mi in items) mi.selection = mi === selectedItem
    }

    private fun updateCuToolchainSelection() {
        val useStorm = Settings.USE_ICESTORM
        iceCube!!.selection = !useStorm
        iceStorm!!.selection = useStorm
    }

    private fun buildHelpMenu() {
        val subMenu = createSubMenu(menu, "Help")
        createItem(subMenu, "About", createSelectionAdapter {
            WelcomeDialog(MainWindow.shell).open()
        })
    }
}